dgl_sparse/src/macro.h

```diff
@@ -52,11 +52,36 @@ namespace sparse {
     }                                                           \
   } while (0)
 
+/**
+ * Dispatch according to Value type (either float or long):
+ *
+ * DGL_SPARSE_VAL_TYPE_SWITCH(tensor.dtype(), ValType, {
+ *   // Now ValType is the type corresponding to data type of the tensor.
+ *   // For instance, one can do this for a CPU array:
+ *   ValType *data = static_cast<ValType *>(array.data_ptr());
+ * });
+ */
+#define DGL_SPARSE_VAL_TYPE_SWITCH(dtype, ValType, op, ...)     \
```

Comment:
The switch macros and impl functions are no longer need. You are directly implement `Compact` in `matrix_ops.cc` because all the operators are torch-based.